### PR TITLE
Register type format

### PR DESCRIPTION
**Pull request by happytree718**
### Related Issue

This PR resolves Issue #77 

### Description

I have implemented a `register` function to set custom type formatting functions, and a `unregister` function to remove it.
A decorator `register_type(obj_type)` is also implemented.
Unit tests are included.
Readme is updated.

### CODE REVIEW

**Comment by gaogaotiantian** [comment]
The overall direction is correct. A couple of suggestions.

1. `register` is a too generic name(I know I used it in the issue, but I did not mean to directy use it as the function name), consider `register_formatter`.
2. Having a decorator is a great idea, but let's combine the two functions :). We can have a single `register_formatter` function for both cases:
  * If only `type` is given, it's a decorator and should return a wrapper to decorate the function
  * If a function is given, it's the inline registry and we just register the function.
3. Using `hex` is a good example, but we can also add an example with `lambda`, which could also be very common and useful.
4. Have you considered subclassing? If I registered a formatting function for the base class, what happens if I'm trying to print a derived class? I'd suggest that we have two changes to address this:
  * Have an optional and keyword only argument `inherit=True`, which indicates whether this formatter would be inherited by subclasses.
  * For each object, go through it's ``__mro__`` to check whether it's parent classes are in the formatter dict.
  * To avoid unnecessary performance hit on common cases, check whether the formatter dict is empty first and shortcut if it is.

**Comment by happytree718** [comment]
1. The reason to use `register` and `unregister` in the first place is that it aligns with the naming in `atexit`. But I will update the names.
2. Will do.
3. Will do.
4. This is something I overlooked. It is a great feature to have, but I have details to clarify:
- For the `inherit` argument, what is the best default value to set it with? I think both `True` and `False` make sense, but which one would you prefer?
- If it is just checking whether the given object is a subclass for one of the types registered, would `issubclass` function be a  better fit? Or does `__mro__` contains any other benefits?

**Comment by gaogaotiantian** [comment]
`atexit` is a more focused library and has less ambiguity for `register`. `objprint` on the other hand, could `register` multiple things, like "a callback after print" or "an output channel to print". So to avoid the ambiguity, we should be clear about what the user is registering.

Yes both `True` and `False` make sense, I'm inclined to make it `True` by default, I think that's more intuitive - subclasses inheriting the feature of the base class.

Yes, the reason we need to go through the `__mro__` is to have a well-defined behavior regardless of the order of the formatter registration. We will always try to find the closest class registered in mro order and that's what Python does.

File: README.md [hunk 1]
```diff
@@ -285,6 +285,73 @@ that match exclusive check.
 
 ```attr_pattern```, ```include``` and ```exclude``` arguments work on ```objprint```, ```objstr``` and ```@add_objprint```.
 
+### Register Custom Type Formatter
+
+You can also customize how certain types of objects are displayed by registering a custom formatter function to transform an object of a specific type into a string. For example, you can print all integers in hexadecimal format by registering the ```hex()``` function for the ```int``` data type, or registering a custom ```lambda``` function. You can always unregister the formatter function if you no longer want to use it with ```unregister_formatter()```, it supports unregistering multiple data types at the same time. And if no input argument is provided to the function, it will clear all the registered functions.
```
Comment (gaogaotiantian) [inline README.md:290]: Please break the lines :) You can have about 100 characters per line(does not have to be precise), but don't write a super long line.

File: README.md [hunk 2]
```diff
@@ -285,6 +285,73 @@ that match exclusive check.
...
+```python
+@op.register_formatter(str)
+def custom_formatter(obj: str):
+    return f"custom_print: {obj}"
+
+# This would clear all the registered formatters 
+op.unregister_formatter()
+```
+
+```Objprint``` always prioritizes custom formatters over default formatters and will use the registered functions if they exist. If type annotations are provided in the custom formatter functions, ```Objprint``` will also perform a validation check to ensure the provided function accepts the specified data type as input and returns a string.
```
Comment (gaogaotiantian) [inline README.md:315]: Use ```objprint```, we never have the capitalized version in our code.

File: README.md [hunk 3]
```diff
@@ -285,6 +285,73 @@ that match exclusive check.
...
+
+The ```op``` function call accepts an ```inherit``` argument for situations involving class inheritance. By setting ```inherit``` to ```True```, a formatter registered for a base class will also be applied to its derived classes, if the formatters for the derived classes are not registered. If a derived class inherits from multiple base classes with registered formatters, the choice of formatter follows the Method Resolution Order (MRO) of the derived class.
```
Comment (gaogaotiantian) [inline README.md:317]: Sorry I may not be clear on this. the ``inherit`` argument is not for ``op``, it's for ``register_formatter``. The user decides whether the formatter should be inherited when they register it,

File: README.md [hunk 4]
```diff
@@ -285,6 +285,73 @@ that match exclusive check.
...
+op.list_formatter()
+```
+
+```
+{
+  BaseClass : base_formatter()
+}
+```
+
+Please note that registering a formatter function with ```op``` will affect the output of ```objprint``` and ```objstr``` methods in the same way. The changes will persist until the formatter is unregistered or ```Objprint``` is reset.
```
Comment (gaogaotiantian) [inline README.md:353]: also change the capitalized version here.

File: src/objprint/objprint.py [hunk 5]
```diff
@@ -130,12 +132,16 @@ def objstr(self, obj: Any, **kwargs) -> str:
...
+            for cls in obj_type.__mro__:
+                if cls in self.type_formatter:
+                    return self.type_formatter[cls](obj)
         # If it's builtin type, return it directly
         if isinstance(obj, str):
             return f"'{obj}'"
-        elif isinstance(obj, int) or \
-                isinstance(obj, float) or \
-                obj is None:
+        elif isinstance(obj, (int, float)) or obj is None:
```
Comment (gaogaotiantian) [inline src/objprint/objprint.py:144]: This is the correct way to check types, but it's note related to this PR. To avoid confusion, we should probably submit this change in another PR(and maybe fix all occurences this kind).

File: src/objprint/objprint.py [hunk 6]
```diff
@@ -130,12 +132,16 @@ def objstr(self, obj: Any, **kwargs) -> str:
         return self._objstr(obj, memo, indent_level=0, cfg=cfg)
 
     def _objstr(self, obj: Any, memo: Optional[Set[int]], indent_level: int, cfg: _PrintConfig) -> str:
+        # If a custom formatter is registered for the object's type, use it directly
+        if cfg.inherit and self.type_formatter:
```
Comment (gaogaotiantian) [inline src/objprint/objprint.py:136]: To have the ``inherit`` option in the formatter itself, you need a data structure like ``{int: {"formatter": formatter, "inherit": True}}``. Check whether the exact type has a formatter first, then go through the mro and see if a base type has the formatter and inherit is True.

File: src/objprint/objprint.py [hunk 7]
```diff
@@ -279,6 +285,46 @@ def add_indent(
...
+
+    def unregister_formatter(self, *obj_types: Type[Any]) -> None:
+        if not obj_types:
+            self.type_formatter.clear()
+        else:
+            for obj_type in obj_types:
+                if obj_type in self.type_formatter:
+                    del self.type_formatter[obj_type]
+
+    def list_formatter(self, print_list: bool = True) -> Optional[dict]:
```
Comment (gaogaotiantian) [inline src/objprint/objprint.py:310]: Having a function to list all the formatters is good, but returning a formatted string is not a great idea. A common use would be to check whether a type already has a formatter. Just return the dictionary would be fine. It's a well-defined format and the user can do whatever they want with the dict.

File: src/objprint/objprint.py [hunk 8]
```diff
@@ -337,3 +383,27 @@ def _get_pack_str(
         else:
             s = ", ".join(elems)
             return f"{header}{s}{footer}"
+
+    def _validate_formatter(
```
Comment (gaogaotiantian) [inline src/objprint/objprint.py:387]: Hmm, I don't think we need this validator. The validator relies on the type annotation and I tend to avoid having run-time check on type annotation for this library. The user should be responsible for their code and we just generate an exception when the code is bad.

**Review by gaogaotiantian — CHANGES_REQUESTED** [review]

File: src/objprint/objprint.py (hunk 8 above)
Comment (happytree718) [reply src/objprint/objprint.py:387]: Sure, I may still keep the obj_type check as it is not relying on type annotation, and is a relatively important property to validate.

File: src/objprint/objprint.py (hunk 7 above)
Comment (happytree718) [reply src/objprint/objprint.py:310]: What the function returns is the dictionary. It is printing the formatted string because the type stored is something like `<class 'int'>`, and I thought it would be more clear to just print `int`. I think it make sense to use `self._objstr()` to get the string and print it out.

File: src/objprint/objprint.py (hunk 5 above)
Comment (happytree718) [reply src/objprint/objprint.py:144]: Acknowledged.

File: src/objprint/objprint.py (hunk 7 above)
Comment (gaogaotiantian) [reply src/objprint/objprint.py:310]: Don't print things if not explicitly issued from the users. I'd suggest to change it to `get_formatters()` and just return a dict. Why would it return None? The user can print with a specific format if they want, but printing the dict itself is clear enough for simple usages. It's not `objprint`'s duty to format this formatter dict well.

File: src/objprint/objprint.py (hunk 6 above)
Comment (happytree718) [reply src/objprint/objprint.py:136]: I would prefer to use a namedtuple to store formatter and inherit, instead of a nested dictionary.

**Comment by happytree718** [comment]
I have made the requested changes. Here is a brief overview:

1. The `inherit` argument is moved to the `register_formatter()`, and I have created a custom namedtuple `FormatterInfo` to store the formatter and its inherit status. Corresponding condition checkings are also updated.
2. The `validate_formatter()` has been removed, but the `obj_type` checking is kept so that an exception will raise when invalid types are used as arguments.
3.  Some code refactoring is reversed and will be changed through a separate PR.
4. Readme and tests are updated correspondingly.

I actually have some other thoughts regarding this feature that I would like to discuss:
Do you think it is useful to allow users to **import a dictionary of formatters**? 
For example, since users can get currently registered formatters by `fmt = op.get_formatter()`, they can store it and import it to a different scope using something like **`op.import_formatter(fmt)`** to reuse these registered formatters. 
This can be helpful when there needs to be a lot of custom formatters that are being repeatedly used in various files.
Is this a desirable feature to have?
If so, then should I make `FormatterInfo` available for import?

Please take a look when you can and let me know if anything else is needed. 

Thank you for your time!

**Comment by gaogaotiantian** [comment]
`import_formatter()` is not helpful because formatters are already registered globally (process-level). It's not trivial to store the formatters on disk to import later. So, the current status is great.
